tests/Unit/ActionTesting.hpp

```diff
@@ -61,6 +61,238 @@ struct get_array_index;
+ *
+ * \snippet Test_ActionTesting.cpp threaded action replace
+ *
+ * Furthermore, simple actions invoked from an action are not run immediately.
+ * Instead, they are queued so that order randomization and introspection may
+ * occur. The simplest form of introspection is checking whether the simple
+ * action queue is empty:
+ *
+ * \snippet Test_ActionTesting.cpp simple action queue empty
+ *
+ * The `ActionTesting::invoke_queued_simple_action()` invokes the next queued
+ * simple action:
+ *
+ * \snippet Test_ActionTesting.cpp invoke queued simple action
+ *
+ * Note that the same functionality exists for threaded actions. The functions
+ * are `ActionTesting::is_threaded_action_queue_empty()`, and
+ * `ActionTesting::invoke_queued_threaded_action()`.
+ *
+ * ### Mocking or replacing components with stubs
+ *
+ * An action can invoke an action on another parallel component. In this case we
+ * need to be able to tell the mocking framework to replace the component the
+ * action is trying to call and instead call a different component that we have
+ * set up to mock that one. For example, the action below invokes an action on
+ * `ComponentB`
+ *
+ * \snippet Test_ActionTesting.cpp action call component b
+ *
+ * Let us assume we cannot using the component `ComponentB` in our test and that
+ * we need to mock it. We do so using the type alias `component_being_mocked` in
+ * the mock component:
+ *
+ * \snippet Test_ActionTesting.cpp mock component b
+ *
+ * When creating the runner `ComponentBMock` is emplaced:
+ *
+ * \snippet Test_ActionTesting.cpp initialize component b
+ *
+ * Any checks and function calls into the ATF also use `ComponentBMock`. For
+ * example:
+ *
+ * \snippet Test_ActionTesting.cpp component b mock checks
+ *
+ * ### Const global cache tags
+ *
+ * Actions sometimes need tags/items to be placed into the `ConstGlobalCache`.
+ * Once the list of tags for the global cache has been assembled, the associated
+ * objects need to be inserted. This is done using the constructor of the
+ * `ActionTesting::MockRuntimeSystem`. For example, consider the tags:
+ *
+ * \snippet Test_ActionTesting.cpp tags for const global cache
+ *
+ * These are added into the global cache by, for example, the metavariables:
+ *
+ * \snippet Test_ActionTesting.cpp const global cache metavars
+ *
+ * A constructor of `ActionTesting::MockRuntimeSystem` takes a
+ * `tuples::TaggedTuple` of the const global cache tags. If you know the order
+ * of the tags in the `tuples::TaggedTuple` you can use the constructor without
+ * explicitly specifying the type of the `tuples::TaggedTuple` as follows:
+ *
+ * \snippet Test_ActionTesting.cpp constructor const global cache tags known
+ *
+ * If you do not know the order of the tags but know all the tags that are
+ * present you can use another constructor where you explicitly specify the
+ * `tuples::TaggedTuple` type:
+ *
+ * \snippet Test_ActionTesting.cpp constructor const global cache tags unknown
```

Comment:
Would it be clearer if both tags in the example had different types, say `int` and `double`? If both arguments in the constructor are the same type, it doesn't really make any difference if you know the order of the tags or not, as the code will always compile.